### PR TITLE
Do more Style/translation checking statically

### PR DESCRIPTION
**Pull request by hypotext**
# Description

**Motivation**: Currently Style does not fail if code does not run, just like JS. So if there is some wrong Style code that does not appear in the energy graph or the shape graph, we don’t report errors. Also, errors are currently thrown at eval-time (JS runtime) and mostly unchecked, so they result in cryptic messages from the JS runtime. 

**Proposal**: Check correctness of the blocks while making the translation, as well as the whole translation before evaluating it, and give people better errors ahead of time, that refer to source locations.

**Implementation**: This PR adds more static checking to Style and the translation:

1. Check the Style AST to make sure all shape, objective, constraint, and function names exist.
2. Check the initialized translation to make sure that all user-mentioned paths exist.

This prevents a large class of runtime JS errors.

Related issue/PR: #473 #474 #475
Related notes: 2/11/21 meeting notes

# Implementation strategy and design decisions

Expressions (`Expr`) are folded over recursively to aggregate all errors.

# Examples with steps to reproduce them

See `compiler/Style.test.ts` and `compiler/venn-3d.sty`.

# Checklist

- [x] I have commented my code, particularly in hard-to-understand areas
- [x] My changes generate no new warnings
- [x] New and existing tests pass locally using `yarn test`
- [x] I ran `yarn docs` and there were no errors when generating the HTML site
- [x] My code follows the style guidelines of this project (e.g.: no ESLint warnings)

# Open questions

Questions that require more discussion or to be addressed in future development:

- [ ] Implement further Style/translation checks described in 2/11/21 meeting notes (e.g. typechecking)
- [ ] Fix any errors that are commented out because a different error is thrown (e.g. `NonexistentNameError`)
- [ ] Report better errors and fixes (use `Identifier`s; reverse-map Substance object translation information to the substitution and Style variable that generated it)
- [ ] Test that multiple errors in a Style program work correctly
- [ ] Make `compiler/venn-3d.sty` a test suite

### CODE REVIEW

**Comment by hypotext** [comment]
This is ready for review, lmk what you think @wodeni. I'm hoping to merge this tomorrow (Thurs) night, so it's available before the session with our first domain expert on Fri.

File: packages/core/src/compiler/Style.ts [hunk 1]
```diff
@@ -1640,12 +1662,195 @@ const translateSubstsBlock = (
...
+  ) {
+    throw Error("internal error: expected function");
+  }
+
+  const fnDict = FN_DICT[expr.tag];
+  if (!fnDict) {
+    throw Error("internal error: unexpected tag");
+  }
+
+  const fnNames: string[] = Object.entries(fnDict).map((e) => e[0]); // Names of built-in functions of that kind
```
Comment (wodeni) [inline packages/core/src/compiler/Style.ts:1731]: Can use this one instead: https://lodash.com/docs/4.17.15#keys

File: packages/core/src/compiler/Style.ts [hunk 2]
```diff
@@ -1640,12 +1662,195 @@ const translateSubstsBlock = (
...
+
+  const givenFnName: Identifier = expr.name;
+
+  if (!fnNames.includes(givenFnName.value)) {
+    const fnErrorType = FN_ERR_TYPE[expr.tag];
+    if (!fnErrorType) {
```
Comment (wodeni) [inline packages/core/src/compiler/Style.ts:1737]: Maybe you could avoid excessive checking like this by narrowing down the type of `expr` first to something like `ObjFn | ConstrFn | CompApp`. It's okay if that's not possible.

File: packages/core/src/compiler/Style.ts [hunk 3]
```diff
@@ -1640,12 +1662,195 @@ const translateSubstsBlock = (
...
+  // NOT CHECKED as this requires more advanced env-building work: At path construction time, check that every Substance object exists in the environment of the block + selector, or that it's defined as a local variable
+
+  const res: StyleResults = block.statements.reduce(
+    (acc: StyleResults, stmt: Stmt): StyleResults =>
+      checkLine(selEnv, stmt, acc),
+    emptyErrs()
+  );
+
+  // TODO(errors): Return warnings (non-fatally); currently there are no warnings though
+  if (res.warnings.length > 0) {
```
Comment (wodeni) [inline packages/core/src/compiler/Style.ts:1844]: Looking at the comment: should this one throw, or you are supposed to do something nonfatal here?

File: packages/core/src/compiler/Style.ts [hunk 4]
```diff
@@ -2633,6 +2853,105 @@ export const disambiguateFunctions = (env: Env, subProg: SubProg) => {
...
+      const res: Path[] = findPathsExpr(fexpr.contents.contents);
+      return acc.concat(res);
+    } else {
+      return acc;
+    }
+  } else if (fexpr.tag === "FGPI") {
+    // Get any exprs that the properties are set to
+    const propExprs: Expr[] = Object.entries(fexpr.contents[1])
+      .map((e) => e[1])
+      .filter((e: TagExpr<VarAD>): boolean => e.tag === "OptEval")
```
Comment (wodeni) [inline packages/core/src/compiler/Style.ts:2933]: You can do this in one single lambda function, so TS would know the type of `e` is changed to `OptEval` (in an `if` statement, for instance).

File: packages/core/src/engine/Evaluator.ts [hunk 5]
```diff
@@ -72,79 +73,79 @@ const dummyIdentifier = (name: string): Identifier => {
...
+
+    // Clone translation to use in this top-level call, because it mutates the translation while interpreting the energy function in order to cache/reuse VarAD (computation) results
+    const trans = clone(transWithVarying);
+
+    // Find out all the GPI expressions in the translation
+    const shapeExprs: IFGPI<VarAD>[] = s.shapePaths.map(
+        (p: Path) => findExprSafe(trans, p) as IFGPI<VarAD>
+    );
+
+    console.log("shapePaths", s.shapePaths.map(prettyPrintPath));
```
Comment (wodeni) [inline packages/core/src/engine/Evaluator.ts:106]: Use logger

File: packages/core/src/utils/Error.ts [hunk 6]
```diff
@@ -20,460 +20,504 @@ const {
...
+
+        case "TaggedSubstanceError": {
+            return showError(error.error); // Substance error
+        }
+
+        // --- BEGIN BLOCK STATIC ERRORS
+
+        case "InvalidGPITypeError": {
+            return `Got invalid GPI type ${error.givenType.value}.`;
+            // COMBAK: GPI type suggestions, or just print the list
```
Comment (wodeni) [inline packages/core/src/utils/Error.ts:190]: I think printing out the list of GPI types is fine. Should be easy to do?

**Review by wodeni — CHANGES_REQUESTED** [review]
Overall looks good to me! Only some minor suggestions below.

File: packages/core/src/compiler/Style.ts (hunk 4 above)
Comment (hypotext) [reply packages/core/src/compiler/Style.ts:2933]: Not sure how I would do this in a lambda in this context, as I would need to be doing both a map and a filter in it. But can do this in the future.

File: packages/core/src/compiler/Style.ts (hunk 3 above)
Comment (hypotext) [reply packages/core/src/compiler/Style.ts:1844]: Yeah, it should throw as a reminder to deal better with warnings in the future. There are currently no warnings in the system, so this is fine. But we should think about how to aggregate/accumulate them in the future, as currently there are only errors, which are reported as if everything is fatal.

File: packages/core/src/utils/Error.ts (hunk 6 above)
Comment (hypotext) [reply packages/core/src/utils/Error.ts:190]: Did this, and the properties too

**Comment by hypotext** [comment]
Thanks @wodeni! Addressed changes in f2657d8. LMK when it's good to merge.

File: packages/core/src/compiler/Style.ts (hunk 4 above)
Comment (wodeni) [reply packages/core/src/compiler/Style.ts:2926]: Just saw the `filter`. Don't really know a great alternative :(.

**Review by wodeni — APPROVED** [review]

**Pull request merged**